crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1654,48 +1654,53 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_fstring_expression(&mut self, fstring: &ast::ExprFString) -> Type<'db> {
         let ast::ExprFString { range: _, value } = fstring;
 
+        let mut has_expression = false;
+        let mut concatenated = String::new();
         for part in value {
             match part {
-                ast::FStringPart::Literal(_) => {
-                    // TODO string literal type
+                ast::FStringPart::Literal(literal) => {
+                    concatenated.push_str(&literal.value);
                 }
                 ast::FStringPart::FString(fstring) => {
-                    let ast::FString {
-                        range: _,
-                        elements,
-                        flags: _,
-                    } = fstring;
-                    for element in elements {
-                        self.infer_fstring_element(element);
+                    for element in &fstring.elements {
+                        match element {
+                            ast::FStringElement::Expression(expression) => {
+                                let ty = self.infer_expression(&expression.expression);
+                                match ty {
+                                    Type::BooleanLiteral(_) | Type::IntLiteral(_) => {
+                                        let repr = format!("{}", ty.representation(self.db));
+                                        concatenated.push_str(&repr);
+                                    }
+                                    Type::StringLiteral(literal) => {
+                                        concatenated.push_str(literal.value(self.db));
+                                    }
+                                    _ => {
+                                        has_expression = true;
+                                    }
+                                }
+                            }
+                            ast::FStringElement::Literal(literal) => {
+                                concatenated.push_str(&literal.value);
+                            }
+                        }
                     }
                 }
             }
+            if concatenated.len() > Self::MAX_STRING_LITERAL_SIZE {
+                break;
```

Comment:
Can't we also break early as soon as we have `has_expression` true?

Reply:
We can - I think I had this in a version and took it out by mistake

Reply:
Ah, good point! Yes, we should infer a type for all expressions, so we need to continue the loop. But we can set a `done` flag so we don't bother looking at the type for subsequent parts, or concatenating anything.